File: binary_indexed_tree/java/TestBinaryIndexedTree.java
import java.util.Random;

public class TestBinaryIndexedTree
{
	public static Random myrand = new Random();
	public static void testGetSum()
	{
		final int LENGTH = 100;
        int[] array = new int[LENGTH];
		for(int i = 0; i < LENGTH; i++)
			array[i] = myrand.nextInt(100);
		BinaryIndexedTree bit = new BinaryIndexedTree(array);
		int sum, right;
		for(int cnt = 0; cnt < 100; cnt++)
		{
			right = myrand.nextInt(LENGTH);
			sum = 0;
			for(int i = 0; i <= right; i++)
				sum += array[i];
			assert sum == bit.getSum(right);
		}
	}
	public static void testUpdate()
	{
		final int LENGTH = 100;
        int[] array = new int[LENGTH];
		for(int i = 0; i < LENGTH; i++)
			array[i] = myrand.nextInt(100);
		BinaryIndexedTree bit = new BinaryIndexedTree(array);
		int sum, right;
		int index, delta;
		for(int cnt = 0; cnt < 100; cnt++)
		{
			index = myrand.nextInt(LENGTH);
			delta = myrand.nextInt(LENGTH);
			array[index] += delta;
			bit.update(index, delta);

			right = myrand.nextInt(LENGTH);
			sum = 0;
			for(int i = 0; i <= right; i++)
				sum += array[i];
			assert sum == bit.getSum(right);
		}
	}
	public static void testRange()
	{
		final int LENGTH = 100;
        int[] array = new int[LENGTH];
		for(int i = 0; i < LENGTH; i++)
			array[i] = myrand.nextInt(100);
		BinaryIndexedTree bit = new BinaryIndexedTree(array);
		int sum, left, right;
		for(int cnt = 0; cnt < 100; cnt++)
		{
			left = myrand.nextInt(LENGTH/2);
			right = left + myrand.nextInt(LENGTH/2);
			sum = 0;
			for(int i = left; i <= right; i++)
				sum += array[i];
			assert sum == bit.getRange(left, right);
		}
	}
	public static void main(String[] args)
	{
		testGetSum();
		testUpdate();
		testRange();
	}
}

File: binary_indexed_tree/c/test_binary_indexed_tree.c
#include "binary_indexed_tree.h"

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

BinaryIndexedTree* bit;
const int LENGTH = 100;
int array[LENGTH];

void set_up()
{
	for(int i = 0; i < LENGTH; i++)
		array[i] = rand() % 100;
	bit = bit_init(array, LENGTH);
}

void break_down()
{
	bit->destroy(bit);
	free(bit);
}

void test_get_sum()
{
	set_up();
	int right, sum;
	for(int cnt = 0; cnt < 100; cnt++)
	{
		right = rand() % LENGTH;
		sum = 0;
		for(int i = 0; i <= right; i++)
			sum += array[i];
		if(bit->get_sum(bit, right) != sum)
			printf("Error in test_get_sum: expect %d, got %d",
					sum, bit->get_sum(bit, right));
	}
	break_down();
}

void test_update()
{
	set_up();
	int right, sum;
	int index, delta;
	for(int cnt = 0; cnt < 100; cnt++)
	{
		index = rand() % LENGTH;
		delta = rand() % 100;
		array[index] += delta;
		bit->update(bit, index, delta);
		right = rand() % LENGTH;
		sum = 0;
		for(int i = 0; i <= right; i++)
			sum += array[i];
		assert(bit->get_sum(bit, right) == sum);
	}
	break_down();
}

void test_get_range()
{
	set_up();
	int left, right, sum;
	for(int cnt = 0; cnt < 100; cnt++)
	{
		left = rand() % LENGTH/2;
		right = left + rand() % LENGTH/2;
		sum = 0;
		for(int i = left; i <= right; i++)
			sum += array[i];
		assert(bit->get_range(bit, left, right) == sum);
	}
	break_down();
}

int main()
{
	srand(time(NULL));
	test_get_sum();
	test_update();
	test_get_range();

}
